### PR TITLE
validate cert

### PR DESCRIPTION
**Pull request by mengqiy**

### CODE REVIEW

**Comment by mengqiy** [comment]
cc: @anfernee

File: pkg/admission/cert/writer/certwriter.go [hunk 1]
```diff
@@ -91,8 +96,16 @@ func handleCommon(webhook *admissionregistrationv1beta1.Webhook, ch certReadWrit
 		return err
 	}
 
+	dnsName, err := dnsNameForWebhook(&webhook.ClientConfig)
+	if err != nil {
+		return err
+	}
 	// Recreate the cert if it's invalid.
-	if !validCert(certs) {
+	valid, err := validCert(certs, dnsName, 6*month)
```
Comment (Liujingfang1) [inline pkg/admission/cert/writer/certwriter.go]: Maybe move `6*month` to a constant.

File: pkg/admission/cert/writer/certwriter.go [hunk 2]
```diff
@@ -141,19 +154,37 @@ type certReadWriter interface {
...
 
-func validCert(certs *generator.Artifacts) bool {
-	// TODO:
-	// 1) validate the key and the cert are valid pair e.g. call crypto/tls.X509KeyPair()
-	// 2) validate the cert with the CA cert
-	// 3) validate the cert is for a certain DNSName
-	// e.g.
-	// c, err := tls.X509KeyPair(cert, key)
-	// err := c.Verify(options)
+func validCert(certs *generator.Artifacts, dnsName string, validTime time.Duration) (bool, error) {
```
Comment (Liujingfang1) [inline pkg/admission/cert/writer/certwriter.go]: All the return statements returns error as `nil`. I think you can remove this argument from the return list. Then `validCert` returns only a `bool`

Comment (mengqiy) [inline pkg/admission/cert/writer/certwriter.go]: The linter is complaining about parameter `validTime` not being changed :/
It is handy to have this parameter for testing purpose.

Comment (mengqiy) [reply pkg/admission/cert/writer/certwriter.go]: Good catch

**Comment by mengqiy** [comment]
PTAL

File: pkg/admission/cert/writer/certwriter.go [hunk 3]
```diff
@@ -39,6 +42,8 @@ const (
 	ServerCertName = "cert.pem"
 )
 
+const sixMonths = 6 * 31 * 24 * time.Hour
```
Comment (droot) [inline pkg/admission/cert/writer/certwriter.go]: does it have to be pkg level const ? I see it only being used inside a fn.

File: pkg/admission/cert/writer/certwriter.go [hunk 4]
```diff
@@ -141,18 +151,36 @@ type certReadWriter interface {
...
+		return false
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return false
+	}
+	ops := x509.VerifyOptions{
+		DNSName:     dnsName,
+		Roots:       pool,
+		CurrentTime: time.Now().Add(sixMonths),
```
Comment (anfernee) [inline pkg/admission/cert/writer/certwriter.go]: why 6 months? also it's more like `time.Now().AddDate(0, 1,  0)` [doc](https://golang.org/pkg/time/#Time.AddDate)

Comment (mengqiy) [reply pkg/admission/cert/writer/certwriter.go]: Good point!
Updated

**Comment by mengqiy** [comment]
PTAL

**Pull request merged**